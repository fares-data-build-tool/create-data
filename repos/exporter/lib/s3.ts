import { AWSError, S3 } from 'aws-sdk';
import { PromiseResult } from 'aws-sdk/lib/request';
import { GetObjectOutput, PutObjectOutput } from 'aws-sdk/clients/s3';
const getS3Client = (): S3 => {
    let options: S3.ClientConfiguration = {
        region: 'eu-west-2',
    };

    if (process.env.NODE_ENV === 'development') {
        options = {
            s3ForcePathStyle: true,
            accessKeyId: 'S3RVER',
            secretAccessKey: 'S3RVER',
            endpoint: 'http://127.0.0.1:4566',
            region: 'eu-west-2',
        };
    }

    return new S3(options);
};

const s3 = getS3Client();

export const getS3Object = (key: string, bucketName: string): Promise<PromiseResult<GetObjectOutput, AWSError>> => {
    try {
        return s3.getObject({ Key: key, Bucket: bucketName }).promise();
    } catch (e) {
        throw new Error(`Failed to get object from S3 Bucket: ${bucketName} with key: ${key}`);
    }
};

export const putS3Object = (
    key: string,
    bucketName: string,
    body: string,
): Promise<PromiseResult<PutObjectOutput, AWSError>> => {
    try {
        return s3.putObject({ Key: key, Bucket: bucketName, Body: body }).promise();
    } catch (e) {
        throw new Error(`Failed to put object in S3 Bucket: ${bucketName} with key: ${key}`);
    }
};
