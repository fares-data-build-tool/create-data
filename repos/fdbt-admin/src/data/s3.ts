import { ListObjectsV2Command, ListObjectsV2CommandOutput, S3Client } from '@aws-sdk/client-s3';
import { AWS_REGION } from '../constants';
import getEnvironment from '../utils/env';

export const getS3Client = async (): Promise<S3Client> => new S3Client({ region: AWS_REGION });

export const getBucketName = (bucketPrefix: string): string => `${bucketPrefix}-${getEnvironment()}`;

export const listBucketObjects = async (s3: S3Client, bucket: string): Promise<ListObjectsV2CommandOutput['Contents']> => {
    const objects: ListObjectsV2CommandOutput['Contents'] = [];

    const getObjectsWithPaginationToken = async (continuationToken: string | undefined) => {
        const listObjectsResponse = await s3.send(new ListObjectsV2Command( {
            Bucket: bucket,
            ContinuationToken: continuationToken,
        }))

        if (listObjectsResponse.Contents) {
            objects.push(...listObjectsResponse.Contents);

            if (listObjectsResponse.NextContinuationToken) {
                await getObjectsWithPaginationToken(listObjectsResponse.NextContinuationToken);
            }
        }
    };

    await getObjectsWithPaginationToken(undefined);

    return objects;
};
