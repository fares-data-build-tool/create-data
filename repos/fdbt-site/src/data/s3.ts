import AWS from 'aws-sdk';

export interface FareStage {
    stageName: string;
    prices: {
        price: string;
        fareZones: string[];
    }[];
}

export interface UserFareStages {
    fareStages: FareStage[];
}

export interface UserFareZone {
    FareZoneName: string;
    NaptanCodes: string;
    AtcoCodes: string;
}

const getS3Client = (): AWS.S3 => {
    let options = {};

    if (process.env.NODE_ENV === 'development') {
        options = {
            s3ForcePathStyle: true,
            accessKeyId: 'S3RVER',
            secretAccessKey: 'S3RVER',
            endpoint: new AWS.Endpoint('http://localhost:4572'),
        };
    }

    return new AWS.S3(options);
};

const s3 = getS3Client();

export const getUserFareStages = async (uuid: string): Promise<UserFareStages> => {
    if (!process.env.USER_DATA_BUCKET_NAME) {
        throw new Error('Environment variable for validated bucket not set');
    }

    const params = {
        Bucket: process.env.USER_DATA_BUCKET_NAME,
        Key: `${uuid}.json`,
    };
    const response = await s3.getObject(params).promise();
    const dataAsString = response.Body?.toString('utf-8') ?? '';

    return JSON.parse(dataAsString);
};

export const putStringInS3 = async (
    bucketName: string,
    key: string,
    text: string,
    contentType: string,
): Promise<void> => {
    const request: AWS.S3.Types.PutObjectRequest = {
        Bucket: bucketName,
        Key: key,
        Body: Buffer.from(text, 'binary'),
        ContentType: contentType,
    };

    await s3.putObject(request).promise();
};

export const putDataInS3 = async (
    data: UserFareZone[] | UserFareStages | string,
    key: string,
    processed: boolean,
): Promise<void> => {
    let contentType = '';
    let bucketName = '';

    if (!process.env.USER_DATA_BUCKET_NAME || !process.env.RAW_USER_DATA_BUCKET_NAME) {
        throw new Error('Bucket name environment variables not set.');
    }

    if (processed) {
        bucketName = process.env.USER_DATA_BUCKET_NAME;
        contentType = 'application/json; charset=utf-8';
    } else {
        bucketName = process.env.RAW_USER_DATA_BUCKET_NAME;
        contentType = 'text/csv; charset=utf-8';
    }

    await putStringInS3(bucketName, key, JSON.stringify(data), contentType);
};
