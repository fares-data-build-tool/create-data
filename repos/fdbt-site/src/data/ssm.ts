import { GetParameterCommand, SSMClient } from '@aws-sdk/client-ssm';

const ssm = new SSMClient({ region: 'eu-west-2' });

export const getSsmValue = async (name: string, withDecryption?: boolean): Promise<string> => {
    const input = {
        Name: name,
        WithDecryption: withDecryption ? withDecryption : true,
    };
    const value = (await ssm.send(new GetParameterCommand(input))).Parameter?.Value;

    if (!value) {
        throw new Error(`Parameter ${name} was not found`);
    }

    return value;
};
