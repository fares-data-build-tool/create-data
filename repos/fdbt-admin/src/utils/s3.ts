import { S3 } from 'aws-sdk';
import { getS3 } from '../data/s3';

interface S3Client {
    client: S3;
}

const getS3Client = async (): Promise<S3Client> => {
    const client = await getS3();
    return { client };
};

export default getS3Client;
