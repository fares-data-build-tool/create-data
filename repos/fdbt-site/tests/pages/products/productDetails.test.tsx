import { shallow } from 'enzyme';
import * as React from 'react';
import {
    getPassengerTypeNameByIdAndNoc,
    getProductMatchingJsonLinkByProductId,
    getSalesOfferPackageByIdAndNoc,
    getTimeRestrictionByIdAndNoc,
} from '../../../src/data/auroradb';
import { getProductsMatchingJson } from '../../../src/data/s3';
import ProductDetails, { getServerSideProps } from '../../../src/pages/products/productDetails';
import {
    expectedCarnetReturnTicket,
    expectedMultiOperatorGeoZoneTicketWithMultipleProducts,
    expectedPeriodGeoZoneTicketWithMultipleProducts,
    expectedPointToPointPeriodTicket,
    expectedSingleTicket,
    getMockContext,
} from '../../testData/mockData';

jest.mock('../../../src/data/auroradb');
jest.mock('../../../src/data/s3');

describe('myfares pages', () => {
    describe('productDetails', () => {
        it('should render correctly', () => {
            const tree = shallow(
                <ProductDetails
                    productName={'Carnet Return Test'}
                    startDate={'18/10/2021'}
                    endDate={'18/10/2121'}
                    productDetailsElements={[
                        { name: 'Service', content: '19 - STAINING - BLACKPOOL via Victoria Hospital (Main Entrance)' },
                        { name: 'Passenger type', content: 'Adult Test' },
                        { name: 'Time restriction', content: 'N/A' },
                        { name: 'Quantity in bundle', content: '2' },
                        { name: 'Carnet expiry', content: '22 year(s)' },
                        { name: 'Purchase methods', content: ['SOP Test 1', 'SOP Test 2'] },
                        { name: 'Start date', content: '18/10/2021' },
                        { name: 'End date', content: '18/10/2121' },
                    ]}
                />,
            );

            expect(tree).toMatchSnapshot();
        });
    });
    describe('getServerSideProps', () => {
        beforeEach(() => {
            (getProductMatchingJsonLinkByProductId as jest.Mock).mockResolvedValueOnce('path');
            (getPassengerTypeNameByIdAndNoc as jest.Mock).mockResolvedValue('Test Passenger Type');

            (getSalesOfferPackageByIdAndNoc as jest.Mock).mockResolvedValueOnce({
                name: 'SOP 1',
            });
            (getSalesOfferPackageByIdAndNoc as jest.Mock).mockResolvedValueOnce({
                name: 'SOP 2',
            });

            (getTimeRestrictionByIdAndNoc as jest.Mock).mockResolvedValue({
                name: 'Test Time Restriction',
            });
        });

        it('correctly returns the elements which should be displayed on the page for a school single ticket', async () => {
            (getProductsMatchingJson as jest.Mock).mockResolvedValueOnce(expectedSingleTicket);
            const ctx = getMockContext({ query: { productId: '1' } });
            expect(await getServerSideProps(ctx)).toStrictEqual({
                props: {
                    productName: 'Test Passenger Type - Single (school)',
                    startDate: '17/12/2020',
                    endDate: '18/12/2020',
                    productDetailsElements: [
                        { name: 'Service', content: '215 - Worthing - Seaham - Crawley' },
                        { name: 'Passenger type', content: 'Test Passenger Type' },
                        { name: 'Only valid during term time', content: 'Yes' },
                        { name: 'Purchase methods', content: ['SOP 1', 'SOP 2'] },
                        { name: 'Start date', content: '17/12/2020' },
                        { name: 'End date', content: '18/12/2020' },
                    ],
                },
            });
        });

        it('correctly returns the elements which should be displayed on the page for a point to point period ticket', async () => {
            (getProductsMatchingJson as jest.Mock).mockResolvedValueOnce(expectedPointToPointPeriodTicket);
            const ctx = getMockContext({ query: { productId: '1' } });
            expect(await getServerSideProps(ctx)).toStrictEqual({
                props: {
                    productName: 'My product',
                    startDate: '17/12/2020',
                    endDate: '18/12/2020',
                    productDetailsElements: [
                        { name: 'Service', content: '215 - Worthing - Seaham - Crawley' },
                        { name: 'Passenger type', content: 'Test Passenger Type' },
                        { name: 'Time restriction', content: 'Test Time Restriction' },
                        { name: 'Period duration', content: '7 weeks' },
                        { name: 'Product expiry', content: '24 hr' },
                        { name: 'Purchase methods', content: ['SOP 1', 'SOP 2'] },
                        { name: 'Start date', content: '17/12/2020' },
                        { name: 'End date', content: '18/12/2020' },
                    ],
                },
            });
        });

        it('correctly returns the elements which should be displayed on the page for a carnet return ticket', async () => {
            (getProductsMatchingJson as jest.Mock).mockResolvedValueOnce({
                ...expectedCarnetReturnTicket,
                returnPeriodValidity: { amount: '3', typeOfDuration: 'month' },
            });
            const ctx = getMockContext({ query: { productId: '1' } });
            expect(await getServerSideProps(ctx)).toStrictEqual({
                props: {
                    productName: 'Test Return Product',
                    startDate: '17/12/2020',
                    endDate: '18/12/2020',
                    productDetailsElements: [
                        { name: 'Service', content: '215 - Worthing - Seaham - Crawley' },
                        { name: 'Passenger type', content: 'Test Passenger Type' },
                        { name: 'Time restriction', content: 'Test Time Restriction' },
                        { name: 'Quantity in bundle', content: '10' },
                        { name: 'Carnet expiry', content: 'No expiry' },
                        { name: 'Return ticket validity', content: '3 month(s)' },
                        { name: 'Purchase methods', content: ['SOP 1', 'SOP 2'] },
                        { name: 'Start date', content: '17/12/2020' },
                        { name: 'End date', content: '18/12/2020' },
                    ],
                },
            });
        });

        it('correctly returns the elements which should be displayed on the page for a Period GeoZone', async () => {
            (getProductsMatchingJson as jest.Mock).mockResolvedValueOnce({
                ...expectedPeriodGeoZoneTicketWithMultipleProducts,
            });
            const ctx = getMockContext({ query: { productId: '1' } });
            expect(await getServerSideProps(ctx)).toStrictEqual({
                props: {
                    productName: 'Weekly Ticket',
                    startDate: '17/12/2020',
                    endDate: '18/12/2020',
                    productDetailsElements: [
                        { name: 'Passenger type', content: 'Test Passenger Type' },
                        { name: 'Zone', content: 'Green Lane Shops' },
                        { name: 'Time restriction', content: 'Test Time Restriction' },
                        { name: 'Period duration', content: '5 weeks' },
                        { name: 'Product expiry', content: '24 hr' },
                        { name: 'Purchase methods', content: ['SOP 1', 'SOP 2'] },
                        { name: 'Start date', content: '17/12/2020' },
                        { name: 'End date', content: '18/12/2020' },
                    ],
                },
            });
        });

        it('correctly returns the elements which should be displayed on the page for a Multi Operator GeoZone Ticket with Multiple Products', async () => {
            (getProductsMatchingJson as jest.Mock).mockResolvedValueOnce({
                ...expectedMultiOperatorGeoZoneTicketWithMultipleProducts,
            });
            const ctx = getMockContext({ query: { productId: '1' } });
            expect(await getServerSideProps(ctx)).toStrictEqual({
                props: {
                    productName: 'Weekly Ticket',
                    startDate: '17/12/2020',
                    endDate: '18/12/2020',
                    productDetailsElements: [
                        { name: 'Passenger type', content: 'Test Passenger Type' },
                        { name: 'Zone', content: 'Green Lane Shops' },
                        { name: 'Time restriction', content: 'Test Time Restriction' },
                        { name: 'Multi Operator Group', content: 'TEST, MCTR, WBTR, BLAC' },
                        { name: 'Period duration', content: '5 weeks' },
                        { name: 'Product expiry', content: '24 hr' },
                        { name: 'Purchase methods', content: ['SOP 1', 'SOP 2'] },
                        { name: 'Start date', content: '17/12/2020' },
                        { name: 'End date', content: '18/12/2020' },
                    ],
                },
            });
        });
    });
});
