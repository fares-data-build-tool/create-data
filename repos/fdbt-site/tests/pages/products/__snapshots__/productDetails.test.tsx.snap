// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`myfares pages productDetails should render correctly 1`] = `
<TwoThirdsLayout
  description="Product Details page of the Create Fares Data Service"
  errors={Array []}
  title="Product Details - Create Fares Data Service"
>
  <BackButton
    href="/products/pointToPointProducts?serviceId=1"
  />
  <div
    className="dft-flex"
  >
    <h1
      className="govuk-heading-l"
      id="product-name"
    >
      Carnet Return Test
    </h1>
    <button
      className="govuk-link govuk-body align-top button-link govuk-!-margin-left-2"
      onClick={[Function]}
    >
      Edit
    </button>
  </div>
  <div
    className="govuk-hint"
    id="product-status"
  >
    Product status: 
    <strong
      className="govuk-tag govuk-tag--turquoise"
    >
      Active
    </strong>
    <strong
      className="govuk-tag govuk-tag--yellow govuk-!-margin-left-2"
    >
      NEEDS ATTENTION
    </strong>
  </div>
  <dl
    className="govuk-summary-list"
    key="Service"
  >
    <div
      className="govuk-summary-list__row"
      key="Service"
    >
      <dt
        className="govuk-summary-list__key"
      >
        Service
      </dt>
      <dd
        className="govuk-summary-list__value"
      >
        <span
          key="19 - STAINING - BLACKPOOL via Victoria Hospital (Main Entrance)"
        >
          19 - STAINING - BLACKPOOL via Victoria Hospital (Main Entrance)
        </span>
      </dd>
    </div>
  </dl>
  <dl
    className="govuk-summary-list"
    key="Passenger type"
  >
    <div
      className="govuk-summary-list__row"
      key="Passenger type"
    >
      <dt
        className="govuk-summary-list__key"
      >
        Passenger type
      </dt>
      <dd
        className="govuk-summary-list__value"
      >
        <div
          className="dft-flex dft-flex-justify-space-between"
        >
          <span
            key="Adult Test"
          >
            Adult Test
          </span>
          <a
            href="/selectPassengerType"
          >
            Edit
          </a>
        </div>
      </dd>
    </div>
  </dl>
  <dl
    className="govuk-summary-list"
    key="Time restriction"
  >
    <div
      className="govuk-summary-list__row"
      key="Time restriction"
    >
      <dt
        className="govuk-summary-list__key"
      >
        Time restriction
      </dt>
      <dd
        className="govuk-summary-list__value"
      >
        <span
          key="N/A"
        >
          N/A
        </span>
      </dd>
    </div>
  </dl>
  <dl
    className="govuk-summary-list"
    key="Quantity in bundle"
  >
    <div
      className="govuk-summary-list__row"
      key="Quantity in bundle"
    >
      <dt
        className="govuk-summary-list__key"
      >
        Quantity in bundle
      </dt>
      <dd
        className="govuk-summary-list__value"
      >
        <span
          key="2"
        >
          2
        </span>
      </dd>
    </div>
  </dl>
  <dl
    className="govuk-summary-list"
    key="Carnet expiry"
  >
    <div
      className="govuk-summary-list__row"
      key="Carnet expiry"
    >
      <dt
        className="govuk-summary-list__key"
      >
        Carnet expiry
      </dt>
      <dd
        className="govuk-summary-list__value"
      >
        <span
          key="22 year(s)"
        >
          22 year(s)
        </span>
      </dd>
    </div>
  </dl>
  <dl
    className="govuk-summary-list"
    key="Purchase methods"
  >
    <div
      className="govuk-summary-list__row"
      key="Purchase methods"
    >
      <dt
        className="govuk-summary-list__key"
      >
        Purchase methods
      </dt>
      <dd
        className="govuk-summary-list__value"
      >
        <div
          className="dft-flex dft-flex-justify-space-between"
        >
          <span
            key="SOP Test 1"
          >
            SOP Test 1
          </span>
          <span
            key="SOP Test 2"
          >
            SOP Test 2
          </span>
          <a
            href="/selectPurchaseMethods"
          >
            Edit
          </a>
        </div>
      </dd>
    </div>
  </dl>
  <dl
    className="govuk-summary-list"
    key="Start date"
  >
    <div
      className="govuk-summary-list__row"
      key="Start date"
    >
      <dt
        className="govuk-summary-list__key"
      >
        Start date
      </dt>
      <dd
        className="govuk-summary-list__value"
      >
        <div
          className="dft-flex dft-flex-justify-space-between"
        >
          <span
            key="18/10/2021"
          >
            18/10/2021
          </span>
          <a
            href="/productDateInformation"
          >
            Edit
          </a>
        </div>
      </dd>
    </div>
  </dl>
  <dl
    className="govuk-summary-list"
    key="End date"
  >
    <div
      className="govuk-summary-list__row"
      key="End date"
    >
      <dt
        className="govuk-summary-list__key"
      >
        End date
      </dt>
      <dd
        className="govuk-summary-list__value"
      >
        <div
          className="dft-flex dft-flex-justify-space-between"
        >
          <span
            key="18/10/2121"
          >
            18/10/2121
          </span>
          <a
            href="/productDateInformation"
          >
            Edit
          </a>
        </div>
      </dd>
    </div>
  </dl>
</TwoThirdsLayout>
`;

exports[`myfares pages productDetails should render correctly for a copied product 1`] = `
<TwoThirdsLayout
  description="Product Details page of the Create Fares Data Service"
  errors={Array []}
  title="Product Details - Create Fares Data Service"
>
  <BackButton
    href="/products/pointToPointProducts?serviceId=1"
  />
  <InformationSummary
    informationText="This is a copy of the product you selected. Edit one or more of the fields as required."
  />
  <div
    className="dft-flex"
  >
    <h1
      className="govuk-heading-l"
      id="product-name"
    >
      Carnet Return Test
    </h1>
    <button
      className="govuk-link govuk-body align-top button-link govuk-!-margin-left-2"
      onClick={[Function]}
    >
      Edit
    </button>
  </div>
  <div
    className="govuk-hint"
    id="product-status"
  >
    Product status: 
    <strong
      className="govuk-tag govuk-tag--turquoise"
    >
      Active
    </strong>
    <strong
      className="govuk-tag govuk-tag--yellow govuk-!-margin-left-2"
    >
      NEEDS ATTENTION
    </strong>
  </div>
  <dl
    className="govuk-summary-list"
    key="Service"
  >
    <div
      className="govuk-summary-list__row"
      key="Service"
    >
      <dt
        className="govuk-summary-list__key"
      >
        Service
      </dt>
      <dd
        className="govuk-summary-list__value"
      >
        <span
          key="19 - STAINING - BLACKPOOL via Victoria Hospital (Main Entrance)"
        >
          19 - STAINING - BLACKPOOL via Victoria Hospital (Main Entrance)
        </span>
      </dd>
    </div>
  </dl>
  <dl
    className="govuk-summary-list"
    key="Passenger type"
  >
    <div
      className="govuk-summary-list__row"
      key="Passenger type"
    >
      <dt
        className="govuk-summary-list__key"
      >
        Passenger type
      </dt>
      <dd
        className="govuk-summary-list__value"
      >
        <div
          className="dft-flex dft-flex-justify-space-between"
        >
          <span
            key="Adult Test"
          >
            Adult Test
          </span>
          <a
            href="/selectPassengerType"
          >
            Edit
          </a>
        </div>
      </dd>
    </div>
  </dl>
  <dl
    className="govuk-summary-list"
    key="Time restriction"
  >
    <div
      className="govuk-summary-list__row"
      key="Time restriction"
    >
      <dt
        className="govuk-summary-list__key"
      >
        Time restriction
      </dt>
      <dd
        className="govuk-summary-list__value"
      >
        <span
          key="N/A"
        >
          N/A
        </span>
      </dd>
    </div>
  </dl>
  <dl
    className="govuk-summary-list"
    key="Quantity in bundle"
  >
    <div
      className="govuk-summary-list__row"
      key="Quantity in bundle"
    >
      <dt
        className="govuk-summary-list__key"
      >
        Quantity in bundle
      </dt>
      <dd
        className="govuk-summary-list__value"
      >
        <span
          key="2"
        >
          2
        </span>
      </dd>
    </div>
  </dl>
  <dl
    className="govuk-summary-list"
    key="Carnet expiry"
  >
    <div
      className="govuk-summary-list__row"
      key="Carnet expiry"
    >
      <dt
        className="govuk-summary-list__key"
      >
        Carnet expiry
      </dt>
      <dd
        className="govuk-summary-list__value"
      >
        <span
          key="22 year(s)"
        >
          22 year(s)
        </span>
      </dd>
    </div>
  </dl>
  <dl
    className="govuk-summary-list"
    key="Purchase methods"
  >
    <div
      className="govuk-summary-list__row"
      key="Purchase methods"
    >
      <dt
        className="govuk-summary-list__key"
      >
        Purchase methods
      </dt>
      <dd
        className="govuk-summary-list__value"
      >
        <div
          className="dft-flex dft-flex-justify-space-between"
        >
          <span
            key="SOP Test 1"
          >
            SOP Test 1
          </span>
          <span
            key="SOP Test 2"
          >
            SOP Test 2
          </span>
          <a
            href="/selectPurchaseMethods"
          >
            Edit
          </a>
        </div>
      </dd>
    </div>
  </dl>
  <dl
    className="govuk-summary-list"
    key="Start date"
  >
    <div
      className="govuk-summary-list__row"
      key="Start date"
    >
      <dt
        className="govuk-summary-list__key"
      >
        Start date
      </dt>
      <dd
        className="govuk-summary-list__value"
      >
        <div
          className="dft-flex dft-flex-justify-space-between"
        >
          <span
            key="18/10/2021"
          >
            18/10/2021
          </span>
          <a
            href="/productDateInformation"
          >
            Edit
          </a>
        </div>
      </dd>
    </div>
  </dl>
  <dl
    className="govuk-summary-list"
    key="End date"
  >
    <div
      className="govuk-summary-list__row"
      key="End date"
    >
      <dt
        className="govuk-summary-list__key"
      >
        End date
      </dt>
      <dd
        className="govuk-summary-list__value"
      >
        <div
          className="dft-flex dft-flex-justify-space-between"
        >
          <span
            key="18/10/2121"
          >
            18/10/2121
          </span>
          <a
            href="/productDateInformation"
          >
            Edit
          </a>
        </div>
      </dd>
    </div>
  </dl>
</TwoThirdsLayout>
`;

exports[`myfares pages productDetails should render correctly for a fare triangle modified product 1`] = `
<TwoThirdsLayout
  description="Product Details page of the Create Fares Data Service"
  errors={Array []}
  title="Product Details - Create Fares Data Service"
>
  <BackButton
    href="/products/pointToPointProducts?serviceId=1"
  />
  <InformationSummary
    informationText="This is a copy of the product you selected. Edit one or more of the fields as required."
  />
  <div
    className="dft-flex"
  >
    <h1
      className="govuk-heading-l"
      id="product-name"
    >
      Carnet Return Test
    </h1>
    <button
      className="govuk-link govuk-body align-top button-link govuk-!-margin-left-2"
      onClick={[Function]}
    >
      Edit
    </button>
  </div>
  <div
    className="govuk-hint"
    id="product-status"
  >
    Product status: 
    <strong
      className="govuk-tag govuk-tag--turquoise"
    >
      Active
    </strong>
    <strong
      className="govuk-tag govuk-tag--yellow govuk-!-margin-left-2"
    >
      NEEDS ATTENTION
    </strong>
  </div>
  <dl
    className="govuk-summary-list"
    key="Service"
  >
    <div
      className="govuk-summary-list__row"
      key="Service"
    >
      <dt
        className="govuk-summary-list__key"
      >
        Service
      </dt>
      <dd
        className="govuk-summary-list__value"
      >
        <span
          key="19 - STAINING - BLACKPOOL via Victoria Hospital (Main Entrance)"
        >
          19 - STAINING - BLACKPOOL via Victoria Hospital (Main Entrance)
        </span>
      </dd>
    </div>
  </dl>
  <dl
    className="govuk-summary-list"
    key="Passenger type"
  >
    <div
      className="govuk-summary-list__row"
      key="Passenger type"
    >
      <dt
        className="govuk-summary-list__key"
      >
        Passenger type
      </dt>
      <dd
        className="govuk-summary-list__value"
      >
        <div
          className="dft-flex dft-flex-justify-space-between"
        >
          <span
            key="Adult Test"
          >
            Adult Test
          </span>
          <a
            href="/selectPassengerType"
          >
            Edit
          </a>
        </div>
      </dd>
    </div>
  </dl>
  <dl
    className="govuk-summary-list"
    key="Time restriction"
  >
    <div
      className="govuk-summary-list__row"
      key="Time restriction"
    >
      <dt
        className="govuk-summary-list__key"
      >
        Time restriction
      </dt>
      <dd
        className="govuk-summary-list__value"
      >
        <span
          key="N/A"
        >
          N/A
        </span>
      </dd>
    </div>
  </dl>
  <dl
    className="govuk-summary-list"
    key="Quantity in bundle"
  >
    <div
      className="govuk-summary-list__row"
      key="Quantity in bundle"
    >
      <dt
        className="govuk-summary-list__key"
      >
        Quantity in bundle
      </dt>
      <dd
        className="govuk-summary-list__value"
      >
        <span
          key="2"
        >
          2
        </span>
      </dd>
    </div>
  </dl>
  <dl
    className="govuk-summary-list"
    key="Carnet expiry"
  >
    <div
      className="govuk-summary-list__row"
      key="Carnet expiry"
    >
      <dt
        className="govuk-summary-list__key"
      >
        Carnet expiry
      </dt>
      <dd
        className="govuk-summary-list__value"
      >
        <span
          key="22 year(s)"
        >
          22 year(s)
        </span>
      </dd>
    </div>
  </dl>
  <dl
    className="govuk-summary-list"
    key="Purchase methods"
  >
    <div
      className="govuk-summary-list__row"
      key="Purchase methods"
    >
      <dt
        className="govuk-summary-list__key"
      >
        Purchase methods
      </dt>
      <dd
        className="govuk-summary-list__value"
      >
        <div
          className="dft-flex dft-flex-justify-space-between"
        >
          <span
            key="SOP Test 1"
          >
            SOP Test 1
          </span>
          <span
            key="SOP Test 2"
          >
            SOP Test 2
          </span>
          <a
            href="/selectPurchaseMethods"
          >
            Edit
          </a>
        </div>
      </dd>
    </div>
  </dl>
  <dl
    className="govuk-summary-list"
    key="Start date"
  >
    <div
      className="govuk-summary-list__row"
      key="Start date"
    >
      <dt
        className="govuk-summary-list__key"
      >
        Start date
      </dt>
      <dd
        className="govuk-summary-list__value"
      >
        <div
          className="dft-flex dft-flex-justify-space-between"
        >
          <span
            key="18/10/2021"
          >
            18/10/2021
          </span>
          <a
            href="/productDateInformation"
          >
            Edit
          </a>
        </div>
      </dd>
    </div>
  </dl>
  <dl
    className="govuk-summary-list"
    key="End date"
  >
    <div
      className="govuk-summary-list__row"
      key="End date"
    >
      <dt
        className="govuk-summary-list__key"
      >
        End date
      </dt>
      <dd
        className="govuk-summary-list__value"
      >
        <div
          className="dft-flex dft-flex-justify-space-between"
        >
          <span
            key="18/10/2121"
          >
            18/10/2121
          </span>
          <a
            href="/productDateInformation"
          >
            Edit
          </a>
        </div>
      </dd>
    </div>
  </dl>
</TwoThirdsLayout>
`;

exports[`myfares pages productDetails should render correctly while the cannot generate return popup is open 1`] = `
<TwoThirdsLayout
  description="Product Details page of the Create Fares Data Service"
  errors={Array []}
  title="Product Details - Create Fares Data Service"
>
  <BackButton
    href="/products/pointToPointProducts?serviceId=1"
  />
  <div
    className="dft-flex"
  >
    <h1
      className="govuk-heading-l"
      id="product-name"
    >
      Carnet Return Test
    </h1>
    <button
      className="govuk-link govuk-body align-top button-link govuk-!-margin-left-2"
      onClick={[Function]}
    >
      Edit
    </button>
  </div>
  <div
    className="govuk-hint"
    id="product-status"
  >
    Product status: 
    <strong
      className="govuk-tag govuk-tag--turquoise"
    >
      Active
    </strong>
    <strong
      className="govuk-tag govuk-tag--yellow govuk-!-margin-left-2"
    >
      NEEDS ATTENTION
    </strong>
  </div>
  <dl
    className="govuk-summary-list"
    key="Service"
  >
    <div
      className="govuk-summary-list__row"
      key="Service"
    >
      <dt
        className="govuk-summary-list__key"
      >
        Service
      </dt>
      <dd
        className="govuk-summary-list__value"
      >
        <span
          key="19 - STAINING - BLACKPOOL via Victoria Hospital (Main Entrance)"
        >
          19 - STAINING - BLACKPOOL via Victoria Hospital (Main Entrance)
        </span>
      </dd>
    </div>
  </dl>
  <dl
    className="govuk-summary-list"
    key="Passenger type"
  >
    <div
      className="govuk-summary-list__row"
      key="Passenger type"
    >
      <dt
        className="govuk-summary-list__key"
      >
        Passenger type
      </dt>
      <dd
        className="govuk-summary-list__value"
      >
        <div
          className="dft-flex dft-flex-justify-space-between"
        >
          <span
            key="Adult Test"
          >
            Adult Test
          </span>
          <a
            href="/selectPassengerType"
          >
            Edit
          </a>
        </div>
      </dd>
    </div>
  </dl>
  <dl
    className="govuk-summary-list"
    key="Time restriction"
  >
    <div
      className="govuk-summary-list__row"
      key="Time restriction"
    >
      <dt
        className="govuk-summary-list__key"
      >
        Time restriction
      </dt>
      <dd
        className="govuk-summary-list__value"
      >
        <span
          key="N/A"
        >
          N/A
        </span>
      </dd>
    </div>
  </dl>
  <dl
    className="govuk-summary-list"
    key="Quantity in bundle"
  >
    <div
      className="govuk-summary-list__row"
      key="Quantity in bundle"
    >
      <dt
        className="govuk-summary-list__key"
      >
        Quantity in bundle
      </dt>
      <dd
        className="govuk-summary-list__value"
      >
        <span
          key="2"
        >
          2
        </span>
      </dd>
    </div>
  </dl>
  <dl
    className="govuk-summary-list"
    key="Carnet expiry"
  >
    <div
      className="govuk-summary-list__row"
      key="Carnet expiry"
    >
      <dt
        className="govuk-summary-list__key"
      >
        Carnet expiry
      </dt>
      <dd
        className="govuk-summary-list__value"
      >
        <span
          key="22 year(s)"
        >
          22 year(s)
        </span>
      </dd>
    </div>
  </dl>
  <dl
    className="govuk-summary-list"
    key="Purchase methods"
  >
    <div
      className="govuk-summary-list__row"
      key="Purchase methods"
    >
      <dt
        className="govuk-summary-list__key"
      >
        Purchase methods
      </dt>
      <dd
        className="govuk-summary-list__value"
      >
        <span
          key="SOP Test 1"
        >
          SOP Test 1
        </span>
        <span
          key="SOP Test 2"
        >
          SOP Test 2
        </span>
      </dd>
    </div>
  </dl>
  <dl
    className="govuk-summary-list"
    key="Start date"
  >
    <div
      className="govuk-summary-list__row"
      key="Start date"
    >
      <dt
        className="govuk-summary-list__key"
      >
        Start date
      </dt>
      <dd
        className="govuk-summary-list__value"
      >
        <div
          className="dft-flex dft-flex-justify-space-between"
        >
          <span
            key="18/10/2021"
          >
            18/10/2021
          </span>
          <a
            href="/productDateInformation"
          >
            Edit
          </a>
        </div>
      </dd>
    </div>
  </dl>
  <dl
    className="govuk-summary-list"
    key="End date"
  >
    <div
      className="govuk-summary-list__row"
      key="End date"
    >
      <dt
        className="govuk-summary-list__key"
      >
        End date
      </dt>
      <dd
        className="govuk-summary-list__value"
      >
        <div
          className="dft-flex dft-flex-justify-space-between"
        >
          <span
            key="18/10/2121"
          >
            18/10/2121
          </span>
          <a
            href="/productDateInformation"
          >
            Edit
          </a>
        </div>
      </dd>
    </div>
  </dl>
</TwoThirdsLayout>
`;
